use services_state module instead of command

alexxa opened the pull request:
Ansible has a built-in module `service_facts` which returns service state information fact data for various service management utilities. It can be used instead of `command: bash -lc "systemctl status <service>"` + `debug` modules. 

Also, that bash command returns non zero code, which results in error print in red even if the `ignore_errors` flag is used. It may disturb a user unnecessarily. `service_facts` will print it in green. 

Compare two outputs: 
Before this PR:
```
TASK [Stop and disable service firewalld] ********************************************************************************
changed: [localhost]

TASK [Get status of firewalld] *******************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["bash", "-lc", "systemctl status firewalld"], "delta": "0:00:00.016379", "end": "2020-10-14 15:52:59.913780", "msg": "non-zero return code", "rc": 3, "start": "2020-10-14 15:52:59.897401", "stderr": "", "stderr_lines": [], "stdout": "● firewalld.service - firewalld - dynamic firewall daemon\n   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)\n   Active: inactive (dead)\n     Docs: man:firewalld(1)\n\nOct 14 13:23:20 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...\nOct 14 13:23:21 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.\nOct 14 13:53:42 ip-10-0-1-135 systemd[1]: Starting firewalld - dynamic firewall daemon...\nOct 14 13:53:42 ip-10-0-1-135 systemd[1]: Started firewalld - dynamic firewall daemon.\nOct 14 13:53:56 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...\nOct 14 13:53:57 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.\nOct 14 15:52:15 ip-10-0-1-135 systemd[1]: Starting firewalld - dynamic firewall daemon...\nOct 14 15:52:15 ip-10-0-1-135 systemd[1]: Started firewalld - dynamic firewall daemon.\nOct 14 15:52:58 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...\nOct 14 15:52:59 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.", "stdout_lines": ["● firewalld.service - firewalld - dynamic firewall daemon", "   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)", "   Active: inactive (dead)", "     Docs: man:firewalld(1)", "", "Oct 14 13:23:20 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...", "Oct 14 13:23:21 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.", "Oct 14 13:53:42 ip-10-0-1-135 systemd[1]: Starting firewalld - dynamic firewall daemon...", "Oct 14 13:53:42 ip-10-0-1-135 systemd[1]: Started firewalld - dynamic firewall daemon.", "Oct 14 13:53:56 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...", "Oct 14 13:53:57 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.", "Oct 14 15:52:15 ip-10-0-1-135 systemd[1]: Starting firewalld - dynamic firewall daemon...", "Oct 14 15:52:15 ip-10-0-1-135 systemd[1]: Started firewalld - dynamic firewall daemon.", "Oct 14 15:52:58 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...", "Oct 14 15:52:59 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon."]}
...ignoring

TASK [debug] *************************************************************************************************************
ok: [localhost] => {
    "command_result.stdout_lines, command_result.stderr_lines": "([u'\\u25cf firewalld.service - firewalld - dynamic firewall daemon', u'   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)', u'   Active: inactive (dead)', u'     Docs: man:firewalld(1)', u'', u'Oct 14 13:23:20 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...', u'Oct 14 13:23:21 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.', u'Oct 14 13:53:42 ip-10-0-1-135 systemd[1]: Starting firewalld - dynamic firewall daemon...', u'Oct 14 13:53:42 ip-10-0-1-135 systemd[1]: Started firewalld - dynamic firewall daemon.', u'Oct 14 13:53:56 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...', u'Oct 14 13:53:57 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.', u'Oct 14 15:52:15 ip-10-0-1-135 systemd[1]: Starting firewalld - dynamic firewall daemon...', u'Oct 14 15:52:15 ip-10-0-1-135 systemd[1]: Started firewalld - dynamic firewall daemon.', u'Oct 14 15:52:58 ip-10-0-1-135 systemd[1]: Stopping firewalld - dynamic firewall daemon...', u'Oct 14 15:52:59 ip-10-0-1-135 systemd[1]: Stopped firewalld - dynamic firewall daemon.'], [])"
}

```
After this PR:
```
TASK [Collect facts about system services] ******************************************************************
ok: [localhost]

TASK [Get status of firewalld] ************************************************************************************************
ok: [localhost] => {
    "msg": "firewalld is 'inactive' and 'disabled'\n"
}

```

berndfinger commented:
Unfortunately, firewalld might not be contained in ansible_facts.services (probably in case firewalld is not installed). See also https://github.com/berndfinger/sap-hana-preconfigure/blob/c17f90294107f32617fff5d9a23b3648feb5e389/tasks/RedHat/generic/assert-firewall.yml#L7 .

alexxa commented:
@berndfinger I have just added a commit, which covers a scenario when firewalld is not installed. It reports an error when firewalld is not installed, and disabls it when it's installed. I used block/rescue modules to completely skip couple of tasks when firewalld is not installed. It will result in ignored/rescued count change in a play recap. There is also an option to handle it without block/rescue modules, and then there will be a different count of skipped and failed tasks. Please let me know if you prefer the latest option, I will update the commit then. 

when firewalld is not installed:
```
TASK [debug] *********************************************************************************************************
ok: [localhost] => {
    "msg": "imported RedHat/generic/configure-firewall.yml"
}

TASK [Gather package facts] ******************************************************************************************
ok: [localhost]

TASK [Stop and disable service firewalld] ****************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Could not find the requested service firewalld: host"}

TASK [Report if firewalld not installed] *****************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Firewalld is not installed"}
...ignoring

PLAY RECAP ***********************************************************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0    skipped=0    rescued=1    ignored=1 
```

and when it's installed: 
```
TASK [Stop and disable service firewalld] ****************************************************************************
changed: [localhost]

TASK [Collect facts about system services] ***************************************************************************
ok: [localhost]

TASK [Get status of firewalld] ***************************************************************************************
ok: [localhost] => {
    "msg": "firewalld is 'inactive' and 'disabled'\n"
}

PLAY RECAP ***********************************************************************************************************
localhost                  : ok=6    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

berndfinger commented:
Many thanks! I have reviewed your proposals and also compared them with similar code in sap-preconfigure. With the presence of the assert code, maybe these "Get status of ..." plays at all. I also would like to avoid multiple gather_..._facts calls. What do you think?

berndfinger commented:
Development and maintenance of this software has stopped. The role [sap_general_preconfigure](https://github.com/sap-linuxlab/community.sap_install/tree/main/roles/sap_hana_preconfigure) replaces it. I have created slx/community.sap_install/issues/21 to track this issue and pull request. I am closing this issue now. Many thanks for your contribution and ideas!
